File: fake-store-quarkus/src/main/java/org/arielzarate/infrastructure/rest/client/ApiFakeStoreClient.java
package org.arielzarate.infrastructure.rest.client;

import org.arielzarate.domain.model.Product;
import org.arielzarate.infrastructure.rest.dto.ProductDto;
import org.arielzarate.infrastructure.rest.mapper.ProductRestMapper;
import org.arielzarate.infrastructure.rest.provider.WebClientMethod;
import org.arielzarate.infrastructure.rest.provider.WebClientProvider;
import jakarta.enterprise.context.ApplicationScoped;

import java.net.URI;
import java.util.Arrays;
import java.util.List;

@ApplicationScoped
public class ApiFakeStoreClient {

    private static final String BASE_URI = "https://fakestoreapi.com";
    private static final long TIMEOUT = 50000L;

    private final WebClientProvider webClientProvider;

    public ApiFakeStoreClient(WebClientProvider webClientProvider) {
        this.webClientProvider = webClientProvider;
    }

    public List<Product> getAllProducts() {
        ProductDto[] response = webClientProvider.applyWithoutBody(
                "FakeStoreAPI",
                WebClientMethod.GET,
                URI.create(BASE_URI + "/products"),
                TIMEOUT,
                null,
                ProductDto[].class
        );

        return Arrays.stream(response)
                .map(ProductRestMapper::toDomain)
                .toList();
    }

    public Product getProductById(Integer id) {
        ProductDto response = webClientProvider.applyWithoutBody(
                "FakeStoreAPI",
                WebClientMethod.GET,
                URI.create(BASE_URI + "/products/" + id),
                TIMEOUT,
                null,
                ProductDto.class
        );

        return ProductRestMapper.toDomain(response);
    }
}

File: api-fake-hexagonal/src/main/kotlin/com/arielZarate/fake_api_hexagonal/infraestructure/rest/ApiFakeStoreClient.kt
package com.arielZarate.fake_api_hexagonal.infraestructure.rest

import com.arielZarate.fake_api_hexagonal.domain.model.Product
import com.arielZarate.fake_api_hexagonal.infraestructure.rest.mapper.ProductMapper
import com.arielZarate.fake_api_hexagonal.infraestructure.rest.model.ProductDto
import com.arielZarate.fake_api_hexagonal.infraestructure.rest.provider.WebClientMethod
import com.arielZarate.fake_api_hexagonal.infraestructure.rest.provider.WebClientProvider
import org.springframework.stereotype.Component
import java.net.URI

@Component
class ApiFakeStoreClient(
    private val webClientProvider: WebClientProvider,
    private val productMapper: ProductMapper
) {

    val urlBase = "https://fakestoreapi.com"

    fun getAllProducts(): List<Product> {
        val response: Array<ProductDto> = webClientProvider.applyWithoutBody(
            clientName = "FakeStoreAPI",
            method = WebClientMethod.GET,
            uri = URI("$urlBase/products"),
            timeout = 10000L,
            headers = null,
            responseTypeReference = Array<ProductDto>::class.java
        )
        return response.toList().map { productMapper.toDomain(it) }
    }

    fun getProductById(id: Int): Product {
        val response: ProductDto = webClientProvider.applyWithoutBody(
            clientName = "FakeStoreAPI",
            method = WebClientMethod.GET,
            uri = URI("$urlBase/products/$id"),
            timeout = 10000L,
            headers = null,
            responseTypeReference = ProductDto::class.java
        )
        return productMapper.toDomain(response)
    }
}